Add rudimentary support for connecting through a ssh tunnel

cosimo opened the pull request:
For Couchbase clusters in certain environments, we don't allow direct access to the rest api admin port (8091, 18091).
Here I tried to implement a ssh-tunnel connection option. When specified, this would automatically open a tunnel to a local port via ssh, and then proceed as usual.

Doing so would allow to setup clusters without having direct access to port 8091.

cosimo-kahoot committed:
Added lightweight wrapper around sshtunnel.SSHTunnelForwarder

cosimo-kahoot committed:
Bugfix: e.printStackTrace() doesn't exist in python (?)

sveniu commented on couchbase_cluster_admin/cluster.py:
```diff
@@ -41,6 +44,25 @@ def __init__(
+            if not ssh_username:
+                raise ValueError("You need to specify a `ssh_username`")
+
+            self.ssh_tunnel = SshTunnel(
```
This doesn't specify local_bind_address, which seems to default to `('0.0.0.0', RANDOM_PORT)`, possibly exposing the endpoint to other clients in the network. It looks like you can specify `('127.0.0.1',)` to get the automatic port.

cosimo replied:
Modified now to use `('127.0.0.1',)` as you suggested.

sveniu commented on couchbase_cluster_admin/cluster.py:
```diff
@@ -41,6 +44,25 @@ def __init__(
+                remote_port=api_port,
+            )
+            self.ssh_tunnel.start()
+            self.api_host = "127.0.0.1"
```
I guess this is always `127.0.0.1`? The api seems to expose both [local_host and local_port](https://github.com/pahaz/sshtunnel/blob/aa7342e55e49529fea020a09ba8f1eed33eb896f/sshtunnel.py#L411-L417) to obain the local endpoint. Dunno how relevant it is with non-127.0.0.1 local hosts..

cosimo-kahoot committed:
Removed mutable list from function argument defaults

cosimo-kahoot committed:
ci.yml: Upgraded poetry to 1.6.1

Getting some weird SolverProblemError failures related to sshtunnel ^0.4.0 ,
Possibly related to https://github.com/python-poetry/poetry/issues/1614 :

```
Creating virtualenv couchbase-cluster-admin-Zcl4Ty5o-py3.8 in /home/runner/.cache/pypoetry/virtualenvs
Installing dependencies from lock file
Warning: The lock file is not up to date with the latest changes in pyproject.toml. You may be getting outdated dependencies. Run update to update them.

  SolverProblemError

  Because couchbase-cluster-admin depends on sshtunnel (^0.4.0) which doesn't match any versions, version solving failed.

  at /opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/poetry/puzzle/solver.py:241 in _solve
      237│             packages = result.packages
      238│         except OverrideNeeded as e:
      239│             return self.solve_in_compatibility_mode(e.overrides, use_latest=use_latest)
      240│         except SolveFailure as e:
    → 241│             raise SolverProblemError(e)
      242│
      243│         results = dict(
      244│             depth_first_search(
      245│                 PackageNode(self._package, packages), aggregate_package_nodes
Error: Process completed with exit code 1.
```

cosimo-kahoot committed:
Refreshed poetry.lock file which is causing the CI failures

cosimo-kahoot committed:
Added python 3.11 to the CI build matrix

cosimo commented:
Rebasing an active PR is against my core values, but I did it anyway... :crossed_fingers:
